Implemented compatibility check for special operations.

mcorino opened the pull request:
* core/src/main/java/org/jacorb/poa/RequestProcessor.java:
      Implemented check for availability of JacORB supplied CORBA compliant
      base classes (through endorsed dirs mechanism) to determin best way
      for invoking special operations.

rnc commented:
Do you have a test to go with this please?

rnc commented on core/src/main/java/org/jacorb/poa/RequestProcessor.java:
```diff
@@ -22,7 +22,10 @@
 
 import java.util.Collection;
 import java.util.HashSet;
+import java.util.Hashtable;
```
Why are we using synchronized hashtable instead of map ?

rnc commented on core/src/main/java/org/jacorb/poa/RequestProcessor.java:
```diff
@@ -361,8 +380,11 @@ else if (servant instanceof org.omg.PortableServer.DynamicImplementation)
                 }
-                if( specialOperations.contains(operation) &&
-                    !(servant instanceof org.jacorb.orb.Forwarder) )
+                SpecialOps special_op = specialOperations.getOrDefault(operation, SpecialOps.NONE);
```
getOrDefault is JDK 8. Currently JacORB is using 1.6 (with a possibility of using 1.7)

rnc commented on core/src/main/java/org/jacorb/poa/RequestProcessor.java:
```diff
@@ -320,9 +336,11 @@ private void invokeOperation()
                 }
 
-                if (specialOperations.contains(operation))
+                SpecialOps special_op = specialOperations.getOrDefault(operation, SpecialOps.NONE);
```
See comment below on getOrDefault

rnc commented on core/src/main/java/org/jacorb/poa/RequestProcessor.java:
```diff
@@ -97,18 +100,31 @@
     private static int count = 0;
 
-    private final static Set<String> specialOperations;
+    private enum SpecialOps { NONE, IS_A, INTERFACE, NON_EXISTENT, GET_POLICY, SET_POLICY, GET_COMPONENT, REPO_ID};
```
Could we not do something like http://www.javapractices.com/topic/TopicAction.do?Id=1 which would simplify this I think ?

mcorino commented:
Ok, I see. I replaced the synchronized set and removed the `getOrDefault()`.
I do not see how any of the mentioned enum practices is going to make this simpler as we need the keys to be Strings and not enums. Using the current map approach I was attempting to avoid having to do string compares more than once (in the `get()`) and be able to simply compare enum instances after that.

I could wip together a simple test that would check the actual operations passed to the DSI servant's `invoke` but setting up the test to run with or without a custom classpath would be one step too far for me as I'm absolutely not familiar with the setup of the JacORB project environment.
